auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -20,6 +43,111 @@ public OIDCAuthCodeImpl(final AuthServiceConfig config) {
      */
     @Override
     public Principal authenticate(final ICredential credential) {
-        throw new IllegalStateException("Not implemented");
+        try {
+            return UserPrincipalImpl
+                    .newUser()
+                    .withAuthenticator(this)
+                    .withUsername(getUsername(credential))
+                    .withCredentials(credential)
+                    .withEmail(getEmail())
+                    .withRoles(getRoles())
+                    .build();
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
+
+    }
+
+    private String getUsername(ICredential credential) throws JSONException {
+        String username = "Unknown Username";
+        userIdentity = getIdentityInformation(credential);
+        if (userIdentity != null) {
+            // get the users username
+            if (userIdentity.has(USERNAME) && userIdentity.getString(USERNAME).length() > 0) {
+                username = userIdentity.getString(USERNAME);
+            }
+        }
+        return username;
+    }
+
+    private String getEmail() throws JSONException {
+        String emailAddress = "Unknown Email";
+        if (userIdentity != null) {
+            // get the users email
+            if (userIdentity.has(EMAIL) && userIdentity.getString(EMAIL).length() > 0) {
+                emailAddress = userIdentity.getString(EMAIL);
+            }
+        }
+        return emailAddress;
+    }
+
+    private List<IRole> getRoles() throws JSONException {
+        List<IRole> roles = new ArrayList();
+        if (userIdentity != null) {
+            List<RealmRole> realmRoles = getRealmRoles();
+            if (realmRoles != null) {
+                roles.addAll(realmRoles);
+            }
+            List<ClientRole> clientRoles = getClientRoles();
+            if (clientRoles != null) {
+                roles.addAll(clientRoles);
+            }
+        }
+        return roles;
+    }
+
+    private List<RealmRole> getRealmRoles() throws JSONException {
+        List realmRoles = new ArrayList<RealmRole>();
+        if (userIdentity.has(REALM) && userIdentity.getJSONObject(REALM).has(ROLES)) {
+            String tokenRealmRolesJSON = userIdentity.getJSONObject(REALM).getString(ROLES);
+
+            String realmRolesString = tokenRealmRolesJSON.substring(1, tokenRealmRolesJSON.length() - 1); //remove the array brackets
+            String roles[] = realmRolesString.split(COMMA);
+            realmRoles = Arrays.asList(roles);
+        }
+        return realmRoles;
+    }
+
+    private List<ClientRole> getClientRoles() throws JSONException {
```

Comment:
Should we be doing all this parsing at "build time"?

To me, this should be done on build and the `clientRoles` and `realmRoles` variables set, and `getClientRoles` and `getRealmRoles` should just return those values instead of performing the parsing here? 

It's a bit confusing when a getter throws a JSONException, to me. wdyt?

Reply:
Edit: I was misreading this block of code